### ISSUE TITLE
X509_verify_cert() behavior change between alpha13 and alpha14

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#14813]
I have a test that does the following

```
ver_ctx = X509_STORE_CTX_new();
store = X509_STORE_new();
X509_STORE_add_cert(store, certificate);
X509_STORE_CTX_init(ver_ctx, store, certificate, NULL);

TS_ASSERT_EQUALS(1, X509_verify_cert(ver_ctx));
```

The 'certificate' is a self-signed (non-CA) certificate. I don't believe details beyond that to be relevant, but it's generated like so

```
certificate = X509_new();
X509_set_version(certificate,3);
ASN1_INTEGER_set(X509_get_serialNumber(certificate),100);
X509_gmtime_adj(X509_get_notBefore(certificate),0);
X509_gmtime_adj(X509_get_notAfter(certificate),(long)60 * 60 * 24 * 365);
X509_set_pubkey(certificate, public_key);
X509_sign(certificate, private_key, md);
```

The 'private_key' and 'public_key' are a key-pair generated elsewhere. 'md' doesn't seem to matter, it fails for various algorithms (SHA1, SHA256, etc.)

Beginning in alpha14, the TS_ASSERT_EQUALS (the meaning of which I hope is obvious even if it's a macro from something not understood by OpenSSL proper) fails.

Interestingly, a similar test assertion that uses X509_verify() continues to work fine.

While I would certainly understand if any deliberate attempt had been made to make such self-signed certificates harder to trust unintentionally, I would worry about the behavior change surprising existing applications. I would have expected a prominent entry in CHANGES if this change in behavior were indeed intentional.

**Comment by t8m** [comment]
@DDvO Any idea why this does not work anymore? If the self-signed certificate is in the trust store, and it is the leaf certificate it should be verified correctly. It seems we regressed on that with some of the recent changes.

X509_verify() just verifies the signature so it is not surprising that it still works.

**Comment by DDvO** [comment]
> The 'certificate' is a self-signed (non-CA) certificate. I don't believe details beyond that to be relevant, 

It turns out they are relevant, though they should not be.

> but it's generated like so
> 
> ```
> certificate = X509_new();
> X509_set_version(certificate,3);

Your cert has X.509 version 4 😛 but this does not matter here.

> ASN1_INTEGER_set(X509_get_serialNumber(certificate),100);
> X509_gmtime_adj(X509_get_notBefore(certificate),0);
> X509_gmtime_adj(X509_get_notAfter(certificate),(long)60 * 60 * 24 * 365);
> X509_set_pubkey(certificate, public_key);
> X509_sign(certificate, private_key, md);
> ```

Your cert has the default subject and issuer name (NULL-DNs), and this is causing trouble.
When building the trivial chain, it needs to be determined that the cert is self-signed (or at least self-issued),
which is implemented using `X509_NAME_cmp(subject, issuer)`, and this currently fails for NULL-DNs.

This function got changed recently in an unfortunate way by #14357, which needs to be fixed.
See https://github.com/openssl/openssl/pull/14589#discussion_r610892756 for details.

**Comment by jwalch** [comment]
Thanks for the excellent triage, @DDvO. I'm glad I included details that I merely suspected were superfluous!

**Comment by DDvO** [comment]
Fix is in #14819.

Until available, you can use as a workaround:
```
        X509_NAME *name = X509_NAME_new();
        X509_NAME_add_entry_by_NID(name, NID_commonName, MBSTRING_ASC,
                                   (unsigned char *)"", -1, -1, 1);
        X509_set_subject_name(certificate, name);
        X509_set_issuer_name(certificate, name);
```
